Improve memory usage of HttpClient's StreamToStreamCopy

stephentoub opened the pull request:
Copying between streams represents one of the larger sources of memory allocations and copying in HttpClient. HttpClient has its own stream-to-stream copy implementation, but there are ways we can significantly improve it based on typical usage within HttpClient.  

This commit does the following:
- Previously it would special case when both the source and destination stream were MemoryStreams, but even in that case it would allocate a temporary buffer and do a Read/Write loop to copy between them, and even then it's relatively rare to copy from a MemoryStream to a MemoryStream.  Instead, if the source stream is a MemoryStream and we can get its buffer, we can do a single write to the destination (regardless of what kind of stream it is), avoiding the need for multiple read/writes and for the temporary buffer.  If we can't get its buffer but its length is known to be less than or equal to the requested buffer size and it's positioned at the beginning, we can simply ToArray the source and still do a single write to the target.
- If the destination is the special LimitMemoryStream, used for buffering output, and if that LimitMemoryStream was pre-sized based on the response having a Content-Length, then rather than allocating a temporary buffer and repeatedly copying from the source stream into that buffer and copying from the buffer into the memory stream, we can instead give the destination's buffer directly to be read into by the source, avoiding the temporary buffer and the extra copies.  This is a very common case.
- If the source does not need to be disposed after the copy, rather than doing our own copy loop, we can instead ask the source to use its CopyToAsync implementation to copy to the destination stream.  Worst case this will be exactly what we were previously doing, but best case the source has its own CopyToAsync override that's more efficient.
- When constructing a LimitMemoryStream, if the Content-Length isn't known, we were explicitly passing a 0 capacity to the base, which causes it to allocate a new zero-length array; I've added a ctor to avoid that.

As a small example, with this short test:
```C#
var uri = new Uri("http://httpbin.org/get");
var c = new HttpClient();
for (int i = 0; i < 100; i++)
    c.GetAsync(uri).GetAwaiter().GetResult().Dispose();
```
before the change it allocated 1,411,714 bytes, and after the change, it allocated 1,003,702, for an ~30% reduction.  That drop is almost entirely in avoiding needing byte[]s:

I'd appreciate some extra scrutiny on this.
cc: @davidsh, @cipop, @ericeil, @kapilash, @bartonjs, @ianhays

davidsh commented:
We should avoid merging this until after we run the UWP (netcore50) related tests (which are not on GitHub yet) especially the ones related to POST'ing data using non-rewindable streams including WinRT based streams that use either IINputStream or IRandomAccessStream.

See (ProjectK TFS):
src\QA\ToF\tests\FX\Conformance\System.Net.Http\4.0.0.0\PostScenarioUWPTest.cs

davidsh commented on src/System.Net.Http/src/System/Net/Http/StreamToStreamCopy.cs:
```diff
@@ -2,59 +2,266 @@
+        /// <param name="source">The source stream from which to copy.</param>
+        /// <param name="destination">The destination stream to which to copy.</param>
+        /// <param name="bufferSize">The size of the buffer to allocate if one needs to be allocated.</param>
+        /// <param name="disposeSource">Whether to dispose of the source stream after the copy has finished successfully.</param>
```
Are XML comments the 'new' thing to do in CoreFx sources?  Have we updated the coding guidelines documents for this and do we have now have tooling that uses this?

davidsh commented:
Although not part of your current PR, you should modify HttpContent.cs and bring this code (below) along for the ride.  You should remove the ifdefs.  I only ifdef'd it at the time when I added it to minimize the changes I was doing several months ago.  But now that netcore50 files are on GitHub, you should take these buffer related changes and incorporate them into this optimization of HttpContent and stream copying:

```c#
#if NETNative        
        internal void SetBuffer(byte[] buffer, int offset, int count)
        {
            _bufferedContent = new MemoryStream(buffer, offset, count, false, true);
        }
        
        internal bool TryGetBuffer(out ArraySegment<byte> buffer)
        {
            if (_bufferedContent == null)
            {
                return false;
            }
            
            return _bufferedContent.TryGetBuffer(out buffer);
        }
#endif
```

stephentoub commented:
> We should avoid merging this until after we run the UWP (netcore50) related tests (which are not on GitHub yet) especially the ones related to POST'ing data using non-rewindable streams including WinRT based streams that use either IInputStream or IRandomAccessStream.

Agreed.  I'd appreciate it if you could help me by doing that.  We can wait until we think the PR is in a good state to merge, so hopefully it only needs to be done once.

> Although not part of your current PR, you should modify HttpContent.cs and bring this code (below) along for the ride. You should remove the ifdefs.

@davidsh, when I make that change (removing the #if around this section and the two call sites in ByteArrayContent), I get the following failure on Windows... this happens even without my stream copy changes, on top of just what's in master:
```
System.Net.Http.Functional.Tests.HttpClientHandlerTest.PostAsync_Post_ChannelBindingHasExpectedValue [FAIL]
        Assert.NotNull() Failure
        Stack Trace:
           c:\Users\stoub\Source\repos\corefx\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.cs(821,0): at System.Net.Http.Functional.Tests.Ht
  tpClientHandlerTest.<PostAsync_Post_ChannelBindingHasExpectedValue>d__52.MoveNext()
```

davidsh commented:
>I'd appreciate it if you could help me by doing that. We can wait until we think the PR is in a good state to merge, so hopefully it only needs to be done once.

That's fine. We can wait until the PR is ready to go before I run the tests.

>when I make that change (removing the #if around this section and the two call sites in ByteArrayContent), I get the following failure on Windows... this happens even without my stream copy changes, on top of just what's in master:

That is interesting. I will need to investigate that. We don't run that particular test for netcore50 (UWP) because HttpClient for UWP (which uses Windows.Web.Http which uses WinInet) does expose channel binding information at this time.

stephentoub commented:
Thanks, David.  I addressed the feedback, including fixing that failing test (though it raises the question about whether this could be considered a breaking change, and if so, do we care?)

davidsh commented:
>I addressed the feedback, including fixing that failing test (though it raises the question about whether this could be considered a breaking change, and if so, do we care?)

I think you are referring to the change you made here in the test:
![image](https://cloud.githubusercontent.com/assets/7684835/14097458/07f8b358-f526-11e5-8a28-5d5c5f185be4.png)

That was an ok change to make and that test should have used a custom content that was directly derived from `HttpContent` instead of deriving from a buffer-based class like `StringContent`.

The current architecture of passing the `TransportContext` object and hence the Channel Binding Token relies on having a custom content (HttpContent based) which is typically a stream type content and having that custom class override the `SerializeToStreamAsync` method.  WCF components are the primary consumer of the channel binding architecture.  Their content objects are complex push-pull stream oriented and derive directly from the `HttpContent` base class.

I think the optimizations that this PR is doing, by using memory buffers if available, won't break these cases.

stephentoub commented:
> Their content objects are complex push-pull stream oriented and derive directly from the HttpContent base class.

Ok, thanks.  Yeah, with regards to breaks, I was referring to the fact that by adding (un-ifdef'ing) the SetBuffer call to the ByteArrayContent ctor, instances derived from ByteArrayContent will no longer have their SerializeToStreamAsync methods called, due to the content already being buffered.  It sounds like anyone deriving from ByteArrayContent rather than HttpContent directly would need to be ok with that, and hence it's fine.

stephentoub commented:
Test Innerloop Windows_NT Release Build and Test please (hang while restoring packages)

bartonjs commented on src/System.Net.Http/src/System/Net/Http/HttpContent.cs:
```diff
@@ -636,19 +623,22 @@ private static int GetPreambleLength(byte[] data, int dataLength, Encoding encod
             }
```
The `default` case here should take the ArraySegment, since Offset is lost on this call.

stephentoub replied:
Good catch. Fixed. (Which also highlights we separately need some tests of various encodings.)

bartonjs commented on src/System.Net.Http/src/System/Net/Http/StreamToStreamCopy.cs:
```diff
@@ -2,59 +2,266 @@
+                    if (capacity > 0 && destinationLimitStream.MaxSize >= capacity)
+                    {
+                        return CopyAsyncToPreSizedLimitMemoryStream(source, destinationLimitStream, bufferSize, disposeSource);
+                    }
```
I don't understand the if here.  The capacity > 0 I can understand; but why does a relationship between MaxSize and Capacity exist?  Length, perhaps... but Capacity?

stephentoub replied:
CopyAsyncToPreSizedLimitMemoryStream will use whatever space is available in the buffer in the LimitMemoryStream.  If that amount is larger than the permitted max size of the stream, CopyAsyncToPreSizedLimitMemoryStream won't care and will happily write beyond the allowed max.  So we double-check here to ensure that the capacity is within the allowed size.  Theoretically we should never trigger this, because the call sites that construct a LimitMemoryStream verify that the capacity they're passing in is less then MaxSize, but that guarantee isn't baked in to LimitMemoryStream, and as such it seemed more prudent here to simply add the condition than to add an assert.

bartonjs commented on src/System.Net.Http/tests/UnitTests/StreamToStreamCopyTest.cs:
```diff
@@ -0,0 +1,351 @@
+    public class StreamToStreamCopyTest
+    {
+        [Theory]
+        [MemberData(nameof(TwoBooleansWithAdditionalArg), new object[] { new object[] { 256, 8192 } })]
```
What about a non-aligned value?  8231 is a nice prime number, so guaranteed to be a partial read for anything other than a buffersize of 8231.

stephentoub replied:
Sure, I can add that.

bartonjs commented on src/System.Net.Http/tests/UnitTests/StreamToStreamCopyTest.cs:
```diff
@@ -0,0 +1,351 @@
+
+            Assert.Equal(input, destination.ToArray());
+            Assert.Equal(input.Length, destination.Position);
+            Assert.Equal(input.Length, destination.Length);
```
I guess this assert is why you used only aligned values?  Because destination.Length is `>=` in the general case, not necessarily equal.

So this feels like it's depending a bit too much on MemoryStream's implementation.  While I can't come up with any "sensible" change in the EnsureCapacity algorithm that doesn't hit 4096, it still feels like you're testing MemoryStream here, and not the StreamToStream copy method.

stephentoub replied:
> Because destination.Length is >= in the general case, not necessarily equal.

No, the length should match the input length in these tests.

> So this feels like it's depending a bit too much on MemoryStream's implementation

How so?  It's not checking Capacity, it's checking Length, which isn't impacted by the internal workings of MemoryStream's growth algorithm.

bartonjs replied:
> How so? It's not checking Capacity, it's checking Length

Because while it says Length, I read Capacity. (d'oh!).  I started thinking in terms of the underlying array, I think.

bartonjs commented on src/System.Net.Http/tests/UnitTests/StreamToStreamCopyTest.cs:
```diff
@@ -0,0 +1,351 @@
+
+            var destination = new HttpContent.LimitMemoryStream(int.MaxValue);
+
+            await StreamToStreamCopy.CopyAsync(source, destination, 0x1000, disposeSource);
```
Up until now you wrote 4096 in decimal, why go hex here?

stephentoub replied:
Habit.  I can change it.

bartonjs commented on src/System.Net.Http/tests/UnitTests/StreamToStreamCopyTest.cs:
```diff
@@ -0,0 +1,351 @@
+                Assert.Equal(input.Length, source.Position);
+            }
+
+            Assert.Equal(input.Skip(StartingPosition), destination.ToArray());
```
Be careful, Steve, you'll scold you for not using Array.Empty\<byte\>() here :smile:.

bartonjs replied:
All of these tests look rather similar from (e.g.) 94-102; perhaps that should be a utility assertion method that you defer to after doing the setup?

stephentoub replied:
> Be careful, Steve, you'll scold you for not using Array.Empty<byte>() here

Where?

> All of these tests look rather similar from (e.g.) 94-102; perhaps that should be a utility assertion method that you defer to after doing the setup?

There are subtle differences in each, and the helper would likely need to take in most of the values being passed to each assert.  Didn't seem worthwhile.

bartonjs replied:
Ugh, all of my comments seem misaligned.  This was in PositionAtEnd, so input.Skip(StartingPosition=input.Length) => empty sequence.

stephentoub replied:
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/IO/MemoryStream.cs#L536

bartonjs commented on src/System.Net.Http/tests/UnitTests/StreamToStreamCopyTest.cs:
```diff
@@ -0,0 +1,351 @@
+        {
+            byte[] input = CreateByteArray(8192);
+            MemoryStream source = CreateSourceMemoryStream(sourceIsExposable, input);
+            var destination = new HttpContent.LimitMemoryStream(int.MaxValue, 1024);
```
None of these tests used a low Max value.  Should they?  Presumably there's some desired exception at some point?

stephentoub replied:
An exception from LimitMemoryStream's ctor you mean?  These tests aren't focused on verifying the ctors of that stream, which already exists.  Or are you talking about something else?

bartonjs replied:
I mean copying past MaxSize.  Since the new stream copy pokes about with its innards it seems like a valid test to me to ensure that if it threw on too-much-copying before that it still throws now.  (If it didn't throw, then I guess I just misunderstand its purpose)

stephentoub replied:
I see.  Sure, I can add a test for that.

bartonjs commented on src/System.Net.Http/src/System/Net/Http/HttpContent.cs:
```diff
@@ -684,30 +674,41 @@ private static bool TryDetectEncoding(byte[] data, int dataLength, out Encoding
-            private int _maxSize;
+            private readonly int _maxSize;
+
+            public LimitMemoryStream(int maxSize)
```
This is just for testing, right?  Since 0 is a legal value for capacity to MemoryStream..ctor(int), why not just explicitly call the other ctor with 0 capacity?

stephentoub replied:
> This is just for testing, right?

No.

> Since 0 is a legal value for capacity to MemoryStream..ctor(int), why not just explicitly call the other ctor with 0 capacity?

Because that base ctor on MemoryStream does ```new byte[capacity]```, causing an array to be allocated unnecessarily. I contemplated adding a special-case to MemoryStream, but this seems like rare usage and that it would be better to handle it here rather than add a conditional there.

bartonjs replied:
> Because that base ctor on MemoryStream does new byte[capacity]

Isn't that what happens here anyways?

LimitMemoryStream..ctor(int) -> MemoryStream..ctor() (implicit `base()`) -> MemoryStream..ctor(0)?

stephentoub replied:
Well... shoot.  I'm not sure what I was looking at when I verified that MemoryStream.ctor() wasn't doing that, but alas.

stephentoub replied:
https://github.com/dotnet/coreclr/pull/3967

stephentoub commented:
@dotnet-bot test this please

jamesqo commented:
@stephentoub, I'm assuming the changes here supplant the ones I made in #5093?

stephentoub commented:
@davidsh, @bartonjs, I've updated the PR based on the feedback.

@jamesqo, yes, but thank you for your PR, regardless.  To be honest, I forgot about it.  It looks like you'd made some of the same changes.

davidsh commented:
@stephentoub I have no further feedback.  Let me know when everyone else has settled on the feedback. Then, please squash this and I will run TFS UWP tests using the ProjectK branch.  That will take a few hours to sync and run tests.

bartonjs commented:
LGTM (pending the external test validation, of course)

stephentoub commented:
Thanks, @davidsh and @bartonjs.  I squashed.

davidsh commented:
TFS UWP tests pass
LGTM

stephentoub commented:
Thanks for verifying, @davidsh.
